### PR TITLE
chore: add datepicker component

### PR DESCRIPTION
**Pull request by jsartisan**
Added daterpicker component along with other components needed for it like Calendar andTimeField

Datepicker
CleanShot 2024-11-21 at 12 38 20@2x

Calendar
CleanShot 2024-11-21 at 12 38 54@2x

Timefield
CleanShot 2024-11-21 at 12 41 56@2x

/ok-to-test tags="@tag.Anvil"

## Release Notes

- **New Features**
	- Introduced a `Calendar` component for date selection and display.
	- Added a `DatePicker` component for selecting dates and times with enhanced error handling.
	- Launched a `TimeField` component for time input with optional prefix and suffix.
	- Updated `TextField` component replacing the previous `TextInput` for improved usability.

- **Bug Fixes**
	- Enhanced styling and responsiveness of input components.

- **Documentation**
	- Added Storybook stories for `Calendar`, `DatePicker`, and `TimeField` components to showcase functionalities and configurations.

- **Chores**
	- Refactored imports to utilize the new `TextField` component across various widgets.

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11970103158>
> Commit: a1a552cb0bfdc9754341de5db0a6d8b142479083
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11970103158&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Anvil`
> Spec:
> <hr>Fri, 22 Nov 2024 10:01:23 UTC

### CODE REVIEW

File: app/client/packages/design-system/widgets/src/components/Calendar/src/CalendarHeading.tsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  const { children, ...domProps } = props;
+
+  return (
+    <Text {...domProps} color="neutral" ref={ref}>
+      {children}
+    </Text>
+  );
+}
+
+const _CalendarHeading = forwardRef(CalendarHeading);
```
Comment (KelvinOm) [inline app/client/packages/design-system/widgets/src/components/Calendar/src/CalendarHeading.tsx:19]: (nit) You can simplify the export a bit here if you do the same as in our other files.

File: app/client/packages/design-system/widgets/src/components/Calendar/src/styles.module.css [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+.calendar tbody [role="button"][data-selected] {
+  background-color: var(--color-bg-accent);
+  color: var(--color-fg-on-accent);
+}
+
+.calendar tbody [role="button"][data-focus-visible],
+.calendar tbody [role="button"][data-focused] {
+  border-color: var(--color-bd-accent);
```
Comment (KelvinOm) [inline app/client/packages/design-system/widgets/src/components/Calendar/src/styles.module.css:65]: It is unclear where the focus is now if the focus is on the selected date.

Снимок экрана 2024-11-21 в 15 31 14

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  ),
+};
+
+export const Loading: Story = {
+  args: {
+    isLoading: true,
+  },
+};
+
+export const Disabled: Story = {
```
Comment (KelvinOm) [inline app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:61]: Except for the main story, the datapicker does not work anywhere else.

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        .map((size) => (
+          <DatePicker key={size} size={size} />
+        ))}
+    </Flex>
+  ),
+};
+
+export const Loading: Story = {
+  args: {
+    isLoading: true,
```
Comment (KelvinOm) [inline app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:57]: I think we shouldn't be able to change the date during the loading.
Снимок экрана 2024-11-21 в 15 36 57

**Review by KelvinOm — CHANGES_REQUESTED** [review]

File: app/client/packages/design-system/widgets/src/components/Calendar/src/Calendar.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+export const Calendar = <T extends DateValue>(props: CalendarProps<T>) => {
+  return (
+    <HeadlessCalendar {...props} className={styles.calendar}>
+      <Flex alignItems="center" justifyContent="space-between" width="100%">
+        <IconButton icon="chevron-left" slot="previous" variant="ghost" />
+        <CalendarHeading size="subtitle" />
+        <IconButton icon="chevron-right" slot="next" variant="ghost" />
+      </Flex>
+      <HeadlessCalendarGrid>
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Calendar/src/Calendar.tsx:29]: It's the storybook styles that we have applied on `td` and `tr`. :/

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx (hunk 3 above)
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:61]: it's the same issue of storybook that we saw in select PR. I'll check again if there is a workaround for this.

File: app/client/packages/design-system/widgets/src/components/Calendar/src/CalendarHeading.tsx (hunk 1 above)
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Calendar/src/CalendarHeading.tsx:19]: I don't ge it. Did you mean to use like:

```jsx
const CalendarHeading = forwardRef((
  props: TextProps,
  ref: ForwardedRef<HTMLHeadingElement>,
) => {
  [props, ref] = useContextProps(props, ref, HeadingContext);
  const { children, ...domProps } = props;

  return (
    <Text {...domProps} color="neutral" ref={ref}>
      {children}
    </Text>
  );
});

export { CalendarHeading };
```

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx (hunk 4 above)
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:57]: so should we disable when loading is true? User can do it anyway like
```<Datepicke isDisabled={isDisabled || isLoading} />

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx (hunk 3 above)
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:61]: Fixed this.

File: app/client/packages/design-system/widgets/src/components/Calendar/src/styles.module.css (hunk 2 above)
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/Calendar/src/styles.module.css]: Fixed

File: app/client/packages/design-system/widgets/src/components/Calendar/src/Calendar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (KelvinOm) [reply app/client/packages/design-system/widgets/src/components/Calendar/src/Calendar.tsx:29]: Can we override it [here](https://github.com/appsmithorg/appsmith/blob/release/app/client/packages/storybook/.storybook/preview-head.html)?

File: app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx (hunk 4 above)
Comment (KelvinOm) [reply app/client/packages/design-system/widgets/src/components/Datepicker/stories/Datepicker.stories.tsx:57]: I looked at how the textfield works and we can also change the value there during loading. Leave it as it is now.

**Review by KelvinOm — APPROVED** [review]

**Pull request merged**